test/apps/iodemo/io_demo.cc

```diff
@@ -172,48 +225,96 @@ class IoDemoRandom {
         }
     }
 
-    static inline void fill(unsigned &seed, void *buffer, size_t size) {
+    static inline void
+    fill(unsigned &seed, void *buffer, size_t size, bool force_host_memory)
+    {
         size_t body_count = size / sizeof(uint64_t);
         size_t tail_count = size & (sizeof(uint64_t) - 1);
         uint64_t *body    = reinterpret_cast<uint64_t*>(buffer);
         uint8_t *tail     = reinterpret_cast<uint8_t*>(body + body_count);
 
-        fill(seed, body, body_count);
-        fill(seed, tail, tail_count);
+        fill(seed, body, body_count, force_host_memory);
+        fill(seed, tail, tail_count, force_host_memory);
     }
 
     static inline size_t validate(unsigned &seed, const void *buffer,
-                                  size_t size) {
+                                  size_t size, bool force_host_memory)
+    {
         size_t body_count    = size / sizeof(uint64_t);
         size_t tail_count    = size & (sizeof(uint64_t) - 1);
         const uint64_t *body = reinterpret_cast<const uint64_t*>(buffer);
         const uint8_t *tail  = reinterpret_cast<const uint8_t*>(body + body_count);
 
-        size_t err_pos = validate(seed, body, body_count);
+        size_t err_pos = validate(seed, body, body_count, force_host_memory);
         if (err_pos < body_count) {
             return err_pos * sizeof(body[0]);
         }
 
-        err_pos = validate(seed, tail, tail_count);
+        err_pos = validate(seed, tail, tail_count, force_host_memory);
         if (err_pos < tail_count) {
             return (body_count * sizeof(body[0])) + (err_pos * sizeof(tail[0]));
         }
 
         return size;
     }
 
+    static void setMemoryType(ucs_memory_type_t memory_type)
+    {
+        _memory_type = memory_type;
+    }
+
+
 private:
-    template <typename T>
-    static inline void fill(unsigned &seed, T *buffer, size_t count) {
+    template<typename T>
+    static inline void
+    fill(unsigned &seed, T *buffer, size_t count, bool force_host_memory)
+    {
+#ifdef HAVE_CUDA
```

Comment:
maybe add a wrapper `cuda_fill`/`cuda_validate` on top of existing host methods?